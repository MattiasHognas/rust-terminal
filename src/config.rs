use serde::Deserialize;

#[derive(Debug, Deserialize, Clone)]
pub struct AppConfig {
    pub tables: Vec<Vec<TableConfig>>,
}

#[derive(Debug, Deserialize, Clone)]
pub struct TableConfig {
#[allow(dead_code)]
    pub id: String,
#[deny(dead_code)]
    pub table_header: Option<String>,
    pub column_headers: Vec<String>,
    pub column_ratios: Vec<u16>,
    pub max_cell_height: usize,
    pub source: TableSource,
    pub design: Option<Design>,
}

#[derive(Debug, Deserialize, Clone)]
#[serde(tag = "type", rename_all = "lowercase")]
#[allow(dead_code)]
pub enum TableSource {
    Static { data: Vec<Vec<String>> },
    File { path: String },
    Http { url: String, refresh_seconds: Option<u64> },
}
#[deny(dead_code)]

#[derive(Debug, Deserialize, Clone)]
pub struct Design {
    pub border: Option<StyleElement>,
    pub header: Option<StyleElement>,
    pub column: Option<StyleElement>,
    pub cell: Option<StyleElement>,
}

#[derive(Debug, Deserialize, Clone)]
pub struct StyleElement {
    pub color: Option<String>,
}

impl AppConfig {
    #[allow(dead_code)]
    pub fn load_from_file(path: &str) -> Result<Self, Box<dyn std::error::Error>> {
    #[deny(dead_code)]
        let data = std::fs::read_to_string(path)?;
        let config: AppConfig = serde_json::from_str(&data)?;
        Ok(config)
    }
}