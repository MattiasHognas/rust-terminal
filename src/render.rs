use ratatui::prelude::*;
use ratatui::widgets::{Block, Borders, Cell, Row, Table};
use crate::config::{AppConfig, TableConfig, TableSource};

pub fn render_app<B: Backend>(terminal: &mut Terminal<B>, config: &AppConfig) -> Result<(), std::io::Error> {
    terminal.draw(|f| {
        let total_area = f.area();

        let row_chunks = Layout::default()
            .direction(Direction::Vertical)
            .constraints(vec![Constraint::Percentage(100 / config.tables.len() as u16); config.tables.len()])
            .split(total_area);

        for (row_idx, row) in config.tables.iter().enumerate() {
            let col_chunks = Layout::default()
                .direction(Direction::Horizontal)
                .constraints(vec![Constraint::Percentage(100 / row.len() as u16); row.len()])
                .split(row_chunks[row_idx]);

            for (col_idx, table) in row.iter().enumerate() {
                render_table::<B>(f, col_chunks[col_idx], table);
            }
        }
    })?;

    Ok(())
}

fn render_table<B: Backend>(f: &mut Frame<'_>, area: Rect, config: &TableConfig) {
    let headers = Row::new(config.column_headers.clone())
        .style(color_from_design(config, "header"));

    let column_constraints: Vec<Constraint> = config
        .column_ratios
        .iter()
        .map(|p| Constraint::Percentage(*p))
        .collect();

    let rows = match &config.source {
        TableSource::Static { data } => data
            .iter()
            .map(|row| {
                let cells = row.iter().map(|cell| {
                    let wrapped = wrap_and_truncate(cell, config.max_cell_height);
                    Cell::from(wrapped).style(color_from_design(config, "cell"))
                });
                Row::new(cells)
            })
            .collect(),
        _ => vec![],
    };

    let mut table = Table::new(rows, column_constraints)
        .header(headers);

    if let Some(title) = &config.table_header {
        let block = Block::default()
            .title(title.as_str())
            .borders(Borders::ALL)
            .border_style(color_from_design(config, "border"));
        table = table.block(block);
    }

    f.render_widget(table, area);
}

fn color_from_design(config: &TableConfig, element: &str) -> Style {
    use ratatui::style::Style;

    let color = match (element, &config.design) {
        ("border", Some(d)) => &d.border,
        ("header", Some(d)) => &d.header,
        ("column", Some(d)) => &d.column,
        ("cell", Some(d)) => &d.cell,
        _ => &None,
    };

    if let Some(style_elem) = color {
        if let Some(hex) = &style_elem.color {
            if let Ok(parsed) = parse_hex_color(hex) {
                return Style::default().fg(parsed);
            }
        }
    }

    Style::default()
}

fn parse_hex_color(hex: &str) -> Result<Color, ()> {
    let hex = hex.trim_start_matches('#');
    if hex.len() != 6 { return Err(()); }
    if let Ok(rgb) = u32::from_str_radix(hex, 16) {
        let r = ((rgb >> 16) & 0xFF) as u8;
        let g = ((rgb >> 8) & 0xFF) as u8;
        let b = (rgb & 0xFF) as u8;
        return Ok(Color::Rgb(r, g, b));
    }
    Err(())
}

fn wrap_and_truncate(text: &str, max_lines: usize) -> String {
    let lines: Vec<String> = text
        .lines()
        .flat_map(|line| textwrap::wrap(line, 30)) // fixed width wrap
        .map(|w| w.into_owned())
        .collect();

    if lines.len() > max_lines {
        let mut truncated = lines[..max_lines].to_vec();
        if let Some(last) = truncated.last_mut() {
            *last = format!("{}...", last.trim_end_matches('.'));
        }
        truncated.join("\n")
    } else {
        lines.join("\n")
    }
}