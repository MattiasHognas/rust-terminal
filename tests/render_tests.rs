#[path = "../src/config.rs"] mod config;
#[path = "../src/render.rs"] mod render;

#[cfg(test)]
mod tests {
    use crate::config::{AppConfig, TableConfig, TableSource, Design, StyleElement};
    use crate::render::{render_app};
    use ratatui::{backend::TestBackend, Terminal};
    use insta::assert_snapshot;

    #[test]
    fn test_render_app() {
        let backend: TestBackend = TestBackend::new(50, 10);
        let mut terminal: Terminal<TestBackend> = Terminal::new(backend).expect("Failed to create terminal");

        let config = AppConfig {
            tables: vec![
                vec![
                    TableConfig {
                        id: "table1".to_string(),
                        table_header: Some("Table header 1".to_string()),
                        column_headers: vec!["Column header 1".to_string(), "Column header 2".to_string()],
                        column_ratios: vec![50, 50],
                        max_cell_height: 3,
                        design: Some(Design {
                            border: Some(StyleElement { color: Some("#ffffff".to_string()) }),
                            header: Some(StyleElement { color: Some("#ffffff".to_string()) }),
                            column: Some(StyleElement { color: Some("#ffffff".to_string()) }),
                            cell: Some(StyleElement { color: Some("#ffffff".to_string()) }),
                        }),
                        source: TableSource::Static {
                            data: vec![vec!["Cell value 1".to_string(), "Cell value 2".to_string()]],
                        },
                    },
                    TableConfig {
                        id: "table2".to_string(),
                        table_header: Some("Table header 2".to_string()),
                        column_headers: vec!["Column header 1".to_string(), "Column header 2".to_string(), "Column header 3".to_string()],
                        column_ratios: vec![33, 33, 33],
                        max_cell_height: 3,
                        design: Some(Design {
                            border: Some(StyleElement { color: Some("#ffffff".to_string()) }),
                            header: Some(StyleElement { color: Some("#ffffff".to_string()) }),
                            column: Some(StyleElement { color: Some("#ffffff".to_string()) }),
                            cell: Some(StyleElement { color: Some("#ffffff".to_string()) }),
                        }),
                        source: TableSource::Static {
                            data: vec![vec!["Cell value 1".to_string(), "Cell value 2".to_string(), "Cell value 3".to_string()]],
                        },
                    },
                ],
                vec![
                    TableConfig {
                        id: "table3".to_string(),
                        table_header: Some("Table header 3".to_string()),
                        column_headers: vec!["Column header 1".to_string()],
                        column_ratios: vec![100],
                        max_cell_height: 3,
                        design: Some(Design {
                            border: Some(StyleElement { color: Some("#ffffff".to_string()) }),
                            header: Some(StyleElement { color: Some("#ffffff".to_string()) }),
                            column: Some(StyleElement { color: Some("#ffffff".to_string()) }),
                            cell: Some(StyleElement { color: Some("#ffffff".to_string()) }),
                        }),
                        source: TableSource::Static {
                            data: vec![vec!["Cell value 1".to_string()]],
                        },
                    },
                ],
            ]
        };

        render_app(&mut terminal, &config).expect("Failed to render app");
        // terminal.flush().expect("Failed to flush terminal");
        assert_snapshot!(terminal.backend());
    }
}