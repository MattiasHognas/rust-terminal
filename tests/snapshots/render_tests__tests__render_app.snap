---
source: tests/render_tests.rs
assertion_line: 75
expression: terminal.backend()
---
"┌Table header 1─────────┐┌Table header 2─────────┐"
"│Column head Column head││Column  Column  Column │"
"│Cell value  Cell value ││Cell va Cell va Cell va│"
"│                       ││                       │"
"└───────────────────────┘└───────────────────────┘"
"┌Table header 3──────────────────────────────────┐"
"│Column header 1                                 │"
"│Cell value 1                                    │"
"│                                                │"
"└────────────────────────────────────────────────┘"
